File: old-quest-system/userQuestsScript.lua
local UserQuestsScript = {}

-- Script properties are defined here
UserQuestsScript.Properties = {
	-- this is fired from the questScript
	{ name = "onQuestComplete", type = "event" },
	{ name = "onQuestActivate", type = "event" },
	{ name = "onQuestProgress", type = "event" },
}

function UserQuestsScript:LocalInit()
	self.quests = self:GetEntity():FindAllScripts("questScript")
	
	self.db = self:GetEntity().documentStoresScript:GetDb("quest-system")
	self.questRewards = self:GatherRewards()
end

function UserQuestsScript:Init()
	self.quests = self:GetEntity():FindAllScripts("questScript")
	
	self.db = self:GetEntity().documentStoresScript:GetDb("quest-system")
	self.questRewards = self:GatherRewards()
	
	local currDay = math.floor(GetWorld():GetUTCTime() / 86400)
	local metadata = self:GetMetadata():FindOne({})
	
	if not metaData or currDay > metadata.lastDaily then 
		self:GetMetadata():UpdateOne({}, {
			_set = {
				lastDaily = currDay
			}
		}, {
			upsert = true
		})
		
		self:ResetDailies()
	end
end

function UserQuestsScript:ResetDailies()
	for _, quest in ipairs(self.quests) do 
		if quest:IsDaily() then 
			quest:ResetQuest()
		end
	end
end

function UserQuestsScript:GetMetadata()
	if not self.metadata then 
		self.metadata = self:GetEntity().documentStoresScript:GetDb("quest-system-metadata")
	end
	
	return self.metadata
end

function UserQuestsScript:GatherRewards()
	local rewards = {}
	for _, reward in ipairs(self:GetEntity():FindAllScripts("questRewardScript")) do 
		local id = reward.properties.questId
		rewards[id] = rewards[id] or {}
		table.insert(rewards[id], reward)
	end
	return rewards
end

function UserQuestsScript:GetQuestRewards(id)
	return self.questRewards[id] or {}
end

function UserQuestsScript:GetUnclaimedRewards()
	local questDocs = self.db:Find({ 
		hasRewards = true,
		claimedRewards = false
	})
	
	local quests = self:MapRecordsToQuests(questDocs)
	local data = {}
	
	for _, q in ipairs(quests) do 
		table.insert(data, {
			id = q.properties.id, 
			rewards = q:GetRewards()
		})
	end
	
	return data
end

function UserQuestsScript:Score()
	local metadata = self:GetMetadata():FindOne({}) or {}
	return metadata.score or 0
end

function UserQuestsScript:ToggleQuestTracking(id)
	self:FindQuest(id):ToggleTracking()
end

function UserQuestsScript:MapRecordsToQuests(records)
	local quests = {}
	
	for _, record in ipairs(records) do 
		table.insert(quests, self:FindQuest(record._id))
	end
	
	return quests
end

function UserQuestsScript:GetTrackedQuests()
	local data = self.db:Find({ tracking = true })
	
	return self:MapRecordsToQuests(data)
end

function UserQuestsScript:AddScore(amt)
	self:GetMetadata():UpdateOne({
		_inc = {
			score = amt
		}
	}, { upsert = true })
	
	self:GetEntity():AddToLeaderboardValue("quest-system-score", amt)
end

function UserQuestsScript:IsQuestComplete(id)
	return self:FindQuest(id):IsComplete()
end

function UserQuestsScript:IsQuestCompleteForDialog(responseData, id)
	responseData.result = self:IsQuestComplete(id)
end

function UserQuestsScript:RecentlyCompleted()
	local quests = self.db:Find({ isEnd = true }, {
		limit = 3,
		sort = {
			completedAt = 1
		}
	})
	
	local retVal = {}
	for _, questScript in ipairs(self.quests) do 
		for _, quest in pairs(quests) do 
			if quest._id == questScript.properties.id then 
				table.insert(retVal, questScript:ToTable())
			end
		end
	end
	
	return retVal
end

function UserQuestsScript:CompleteQuest(id)
	if not IsServer() then 
		self:SendToServer("CompleteQuest", id) 
		return
	end
	
	local quest = self:FindQuest(id)
	
	if not quest then return end 
	if quest:IsComplete() then return end 
	if not quest:IsActive() then return end
	
	--print("Completing quest")
	quest:Complete()
end

function UserQuestsScript:FindQuest(id)
	for _, quest in ipairs(self.quests) do 
		if quest.properties.id == id then 
			return quest
		end
	end
end

function UserQuestsScript:BuildStructure(quests, structure)
	for _, quest in ipairs(quests) do 
		local s = {}
		structure[quest.properties.id] = s
		self:BuildStructure(quest:ChildQuests(), s)
	end
end

function UserQuestsScript:GetStructure()
	local structure = {}
	for _, quest in ipairs(self.quests) do 
		if quest:IsRoot() then 
			local struct = {}
			structure[quest.properties.id] = struct
			self:BuildStructure(quest:ChildQuests(), struct)
		end
	end
	
	--self:GetEntity().cUtilsScript:PrintTable(structure)
	
	return structure
end

function UserQuestsScript:GetWidgetData()
	local quests = {}
	
	for _, quest in ipairs(self.quests) do 
		table.insert(quests, quest:ToTable())
	end
	
	return quests
end

return UserQuestsScript

File: scripts/user_quests_script.rb
module Hoard
  module Scripts
    class UserQuestsScript < Hoard::Script
      attr_accessor :on_quest_complete, :on_quest_activate, :on_quest_progress
      attr_reader :quests, :quest_rewards

      event :on_quest_complete
      event :on_quest_activate
      event :on_quest_progress
      
      def initialize()
        @quests = []
        @quest_rewards = {}
      end

      def local_init
        gather_quests
        gather_rewards
        build_quest_tree
        reset_dailies_if_needed
      end

      def db
        user&.document_stores_script&.get_db("quest_system")
      end

      def metadata_db
        user&.document_stores_script&.get_db("quest_system_metadata")
      end

      def gather_quests
        # This assumes quests are attached to entities that are children of the User's entity,
        # or some other central entity. For now, we'll search all entities.
        @quests = []
        ObjectSpace.each_object(Hoard::Entity) do |entity|
          entity.find_scripts(Hoard::Scripts::QuestScript).each do |quest_script|
            @quests << quest_script
          end
        end
      end

      def gather_rewards
        @quest_rewards = {}
        ObjectSpace.each_object(Hoard::Entity) do |entity|
          entity.find_scripts(Hoard::Scripts::QuestRewardScript).each do |reward_script|
            @quest_rewards[reward_script.quest_id] ||= []
            @quest_rewards[reward_script.quest_id] << reward_script
          end
        end
      end

      def build_quest_tree
        quests_by_id = @quests.to_h { |q| [q.id, q] }
        @quests.each do |quest|
          if quest.parent_id
            parent_quest = quests_by_id[quest.parent_id]
            if parent_quest
              parent_quest.children << quest
              quest.parent = parent_quest
            end
          end
          if quest.prerequisite
            quest.prerequisite_quest = quests_by_id[quest.prerequisite]
          end
        end
      end

      def reset_dailies_if_needed
        curr_day = (Time.now.utc.to_i / 86400)
        metadata = metadata_db&.find_one({}) || {}
        last_daily = metadata[:last_daily]

        if !last_daily || curr_day > last_daily
          metadata_db&.update_one({}, { "$set" => { last_daily: curr_day } }, { upsert: true })
          reset_dailies
        end
      end

      def reset_dailies
        @quests.each do |quest|
          quest.db&.delete_one({ _id: quest.id }) if quest.daily
        end
      end

      def get_quest_rewards(id)
        @quest_rewards[id] || []
      end

      def get_unclaimed_rewards
        quest_docs = db&.find({ has_rewards: true, claimed_rewards: false }) || []
        quests = map_records_to_quests(quest_docs)
        quests.map do |q|
          {
            id: q.id,
            rewards: q.rewards.map(&:to_h)
          }
        end
      end

      def score
        metadata = metadata_db&.find_one({}) || {}
        metadata[:score] || 0
      end

      def add_score(amount)
        metadata_db&.update_one({}, { "$inc" => { score: amount } }, { upsert: true })
        # TODO: Add to leaderboard
      end

      def toggle_quest_tracking(id)
        find_quest(id)&.toggle_tracking
      end

      def map_records_to_quests(records)
        records.map { |record| find_quest(record[:_id]) }.compact
      end

      def get_tracked_quests
        data = db&.find({ tracking: true }) || []
        map_records_to_quests(data)
      end

      def is_quest_complete?(id)
        find_quest(id)&.is_complete?
      end

      def recently_completed
        quests = db&.find({ is_end: true }, { sort: { completed_at: -1 }, limit: 3 }) || []
        map_records_to_quests(quests).map(&:to_h)
      end

      def complete_quest(id)
        # In Hoard, scripts run on the server, so no need for SendToServer
        quest = find_quest(id)
        return unless quest
        return if quest.is_complete?
        return unless quest.is_active?
        quest.complete
      end

      def find_quest(id)
        @quests.find { |q| q.id == id }
      end

      def get_widget_data
        @quests.map(&:to_h)
      end
    end
  end
end
